architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,677 @@
+1. Post-upgrade verification confirms system health
+1. User is notified about the CLI version mismatch
+
+**Exceptions:**
+
+1. If version check fails because the target is lower than current version
+2. If components fail health checks after upgrade
+
+#### Scenario 2: DevOps engineer upgrades with custom configuration
+
+**User Story:** As a DevOps engineer, I need to upgrade Radius with custom configuration parameters to match organization's infrastructure requirements.
+
+**User Experience:**
+
+```bash
+# Upgrade with custom configuration values
+> rad upgrade kubernetes --version v0.44.0 --set global.monitoring.enabled=true --set global.resources.limits.memory=2Gi
+
+Initiating Radius upgrade from v0.43.0 to v0.44.0 with custom configuration...
+Custom configuration detected:
+  - global.monitoring.enabled: true
+  - global.resources.limits.memory: 2Gi
+Pre-flight checks:
+  ✓ Valid version target
+  ✓ Compatible upgrade path
+  ✓ Custom configuration validated
+Creating backup of current user data...
+  ✓ Backup created successfully
+Upgrading control plane components with custom configuration:
+  ✓ Universal Control Plane
+  ✓ Deployment Engine
+  ✓ Applications Resource Provider
+  ✓ Controller
+  ✓ ...
+Applying custom configuration settings...
+Performing post-upgrade verification...
+  ✓ All components healthy
+  ✓ Custom configuration applied successfully
+
+Upgrade complete! Radius has been successfully upgraded to v0.44.0 with your custom configuration.
+```
+
+**Result:**
+
+1. System acknowledges custom configuration parameters
+1. Pre-flight checks validate both upgrade path and configuration
+1. Control plane components are upgraded with custom settings
+1. Verification confirms both health and configuration application
+
+**Exceptions:**
+
+1. If custom configuration parameters are invalid (or should it be ignored?)
+
+#### Scenario 3: Handling upgrade failure and recovery
+
+**User Story:** As a system administrator, I need confidence that if an upgrade fails, the system can recover without data loss or extended downtime.
+
+**User Experience:**
+
+```bash
+# Attempt upgrade that encounters an issue
+> rad upgrade kubernetes --version v0.44.0
+
+Initiating Radius upgrade from v0.43.0 to v0.44.0...
+Pre-flight checks:
+  ✓ Valid version target
+  ✓ Compatible upgrade path
+Creating backup of current user data...
+  ✓ Backup created successfully
+Upgrading control plane components:
+  ✓ Universal Control Plane
+  ✓ Deployment Engine
+  ✗ Applications Resource Provider (ERROR: Container image pull failed)
+
+ERROR: Upgrade failed during Applications Resource Provider update.
+Initiating automatic rollback to v0.43.0...
+  ✓ Restoring from backup (Not sure if this is needed)
+  ✓ Universal Control Plane reverted
+  ✓ Deployment Engine reverted
+  ✓ System verification complete
+  ✓ ...
+
+Rollback complete. System has been restored to v0.43.0.
+Review Kubernetes events and logs for more details on the failure.
```

Comment:
Do we want the user to switch back and forth between rad cli and kubectl? Can this be achieved through `rad debug-logs`?

Reply:
Probably. The outputs will be updated. These are just placeholders for now.